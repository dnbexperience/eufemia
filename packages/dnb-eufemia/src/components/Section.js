/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index section to autogenerate all the components and patterns
 * Used by "prepareSections"
 */

import Section from './section/Section'
export * from './section/Section'
export default Section
