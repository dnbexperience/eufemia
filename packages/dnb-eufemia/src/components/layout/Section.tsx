import React from 'react'
import classnames from 'classnames'
import FlexContainer, {
  Props as FlexContainerProps,
} from './FlexContainer'
import type { ComponentProps } from '../../extensions/forms/types'

export type Props = ComponentProps & {
  direction?: FlexContainerProps['direction']
  spacing?: FlexContainerProps['spacing']
  children: React.ReactNode
}

function Section(props: Props) {
  const { className, direction, spacing, children, ...rest } = props
  return (
    <FlexContainer
      element="section"
      className={classnames('dnb-layout__section', className)}
      direction={direction ?? 'column'}
      spacing={spacing ?? 'small'}
      {...rest}
    >
      {children}
    </FlexContainer>
  )
}

Section._supportsEufemiaSpacingProps = true
export default Section
