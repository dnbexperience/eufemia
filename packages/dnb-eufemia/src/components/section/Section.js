/**
 * Web Section Component
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import Context from '../../shared/Context'
import {
  isTrue,
  registerElement,
  validateDOMAttributes,
  processChildren,
  extendPropsWithContext,
} from '../../shared/component-helper'
import {
  spacingPropTypes,
  createSpacingClasses,
} from '../space/SpacingHelper'

export default class Section extends React.PureComponent {
  static tagName = 'dnb-section'
  static contextType = Context

  static propTypes = {
    style_type: PropTypes.string,
    spacing: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    element: PropTypes.string,
    inner_ref: PropTypes.object,

    ...spacingPropTypes,

    class: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.func,
      PropTypes.node,
    ]),
  }

  static defaultProps = {
    style_type: null,
    spacing: null,
    element: 'section',
    inner_ref: null,
    class: null,

    className: null,
    children: null,
  }

  static enableWebComponent() {
    registerElement(Section?.tagName, Section, Section.defaultProps)
  }

  static getContent(props) {
    if (props.text) return props.text
    return processChildren(props)
  }

  constructor(props) {
    super(props)
    this._ref = props.inner_ref || React.createRef()
  }

  render() {
    // use only the props from context, who are available here anyway
    const props = extendPropsWithContext(
      this.props,
      Section.defaultProps,
      this.context.FormRow,
      this.context.Section
    )

    const {
      element,
      style_type,
      spacing,
      className,
      class: _className,

      ...attributes
    } = props

    const content = Section.getContent(this.props)

    const params = {
      className: classnames(
        'dnb-section',
        `dnb-section--${style_type || 'mint-green-12'}`,
        (isTrue(spacing) || spacing) &&
          `dnb-section--spacing${
            !/true|false/.test(String(spacing)) ? '-' + spacing : ''
          }`,
        createSpacingClasses(props),
        className,
        _className
      ),
      ...attributes,
    }

    // also used for code markup simulation
    validateDOMAttributes(this.props, params)

    const Element = element || 'section'

    return (
      <Element {...params} ref={this._ref}>
        {content}
      </Element>
    )
  }
}
