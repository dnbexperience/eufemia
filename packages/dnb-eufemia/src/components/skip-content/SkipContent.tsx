import React from 'react'
import classnames from 'classnames'
import Button, { ButtonProps } from '../button/Button'
import HeightAnimation from '../height-animation/HeightAnimation'
import { applyPageFocus } from '../../shared/helpers'

export interface SkipContentProps {
  /**
   * Define an existing `Section` id to focus when the inner button got pressed.
   * Required
   */
  selector: string

  /**
   * Define a clear message describing the choices the users has.
   * Optional
   */
  text?: React.ReactNode

  /**
   * Defines the delay after the enter key has been pressed
   * Defaults to 400
   */
  focusDelay?: number
}

export type SkipContentAllProps = SkipContentProps & ButtonProps

interface FocusEvent<T = Element> extends React.SyntheticEvent<T> {
  target: EventTarget & T
}

const SkipContent = (localProps: SkipContentAllProps) => {
  const {
    selector,
    text,
    children,
    className,
    focusDelay = 400,
    ...props
  } = localProps

  React.useEffect(
    () => () => {
      clearTimeout(timeout.current)
    },
    []
  )

  const [visible, setVisible] = React.useState(false)
  const [keepReturnActive, setKeepReturnActive] = React.useState(false)
  const ref = React.useRef<HTMLElement>()
  const timeout = React.useRef<NodeJS.Timeout>()

  const classes = classnames(
    'dnb-skip-content',
    visible && 'dnb-skip-content--visible',
    keepReturnActive && 'dnb-skip-content__return--active',
    className
  )
  const returnSelector = selector.replace(/^(\.|#)/, '')
  const returnId = `${returnSelector}--alias`

  return (
    <span
      className={classes}
      ref={ref}
      onFocus={handleFocus}
      id={returnId}
    >
      <>
        {!visible && (
          <button className="dnb-sr-only" onKeyUp={handleKeyUp}>
            {text || children}
          </button>
        )}
        <HeightAnimation open={visible}>
          <Button
            wrap
            variant="secondary"
            onClick={handleClick}
            onBlur={handleBlur}
            {...(props as Record<string, unknown>)}
          >
            {text || children}
          </Button>
        </HeightAnimation>
      </>
    </span>
  )

  function handleBlur() {
    setVisible(false)
  }

  function handleClick() {
    setVisible(false)

    // Scroll to the element at first
    const element = document.querySelector(selector)
    element?.scrollIntoView?.({ behavior: 'smooth' })
    element?.classList.add('dnb-skip-content__focus')

    // Delay the focus, so the UX is smoother
    timeout.current = setTimeout(() => {
      applyPageFocus(selector)

      // Tell the linked return component, it should stay active (if it gets focused as well)
      document
        .querySelector(`#${returnSelector}--alias--alias`)
        ?.classList.add('dnb-skip-content__return--active')
    }, focusDelay)
  }

  function handleFocus(e: FocusEvent) {
    if (e.target.tagName === 'SPAN') {
      setFocus()
    }
  }

  function handleKeyUp(e: React.KeyboardEvent) {
    if (e.key === 'Tab') {
      setFocus()
    }
  }

  function setFocus() {
    setVisible(true)

    // Wait one frame, so ref is set
    window.requestAnimationFrame(() => {
      const element = ref.current?.querySelector(
        '.dnb-button'
      ) as HTMLElement
      element?.focus()
    })

    // Ensure the __return button stays active
    if (ref.current?.getAttribute('class').includes('__return--active')) {
      setKeepReturnActive(true)
    }
  }
}

export type SkipContentReturnProps = SkipContentAllProps

const SkipContentReturn = (localProps: SkipContentReturnProps) => {
  const { selector, className, ...props } = localProps

  const classes = classnames('dnb-skip-content__return', className)

  return (
    <SkipContent
      selector={`${selector}--alias`}
      className={classes}
      {...props}
    />
  )
}

SkipContent.Return = SkipContentReturn

SkipContent._supportsSpacingProps = true

export default SkipContent
