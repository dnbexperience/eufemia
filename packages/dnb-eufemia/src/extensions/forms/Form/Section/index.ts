export { default } from './Section'
export * from './Section'
