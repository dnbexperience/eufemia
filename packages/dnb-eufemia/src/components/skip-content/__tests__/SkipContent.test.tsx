import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import SkipContent, { SkipContentAllProps } from '../SkipContent'
import Section from '../../Section'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'

describe('SkipContent', () => {
  it('renders with properties as object', () => {
    const props: SkipContentAllProps = { selector: '#unique-id' }
    render(
      <>
        <SkipContent {...props} />
        <Section id="unique-id">content</Section>
      </>
    )

    expect(document.querySelector('.dnb-skip-content')).toBeInTheDocument()
  })

  it('should have one button child with dnb-sr-only class', () => {
    render(
      <>
        <SkipContent selector="#unique-id" text="Text" />
        <Section id="unique-id">content</Section>
      </>
    )

    const element = document.querySelector('.dnb-skip-content')
    expect(element.childNodes).toHaveLength(1)
    expect((element.childNodes[0] as HTMLButtonElement).className).toBe(
      'dnb-sr-only'
    )
    expect((element.childNodes[0] as HTMLButtonElement).tagName).toBe(
      'BUTTON'
    )
  })

  it('should have no aria-hidden on button', () => {
    render(
      <>
        <SkipContent selector="#unique-id" text="Text" />
        <Section id="unique-id">content</Section>
      </>
    )

    const element = document.querySelector('.dnb-skip-content')
    expect(element.childNodes[0] as HTMLButtonElement).not.toHaveAttribute(
      'aria-hidden'
    )
  })

  it('should have a certain amount of attributes on button', () => {
    render(
      <>
        <SkipContent selector="#unique-id" text="Text" />
        <Section id="unique-id">content</Section>
      </>
    )

    const element = document.querySelector('.dnb-skip-content')
    expect((element.childNodes[0] as HTMLButtonElement).attributes)
      .toMatchInlineSnapshot(`
      NamedNodeMap {
        "class": "dnb-sr-only",
      }
    `)
  })

  it('should have given text in DOM', () => {
    render(
      <>
        <SkipContent
          selector="#unique-id"
          text="Skip table with enter key – or continue tabbing"
        />
        <Section id="unique-id">content</Section>
      </>
    )

    const element = document.querySelector('.dnb-skip-content')
    expect(element.textContent).toBe(
      'Skip table with enter key – or continue tabbing'
    )
  })

  it('should show button and set focus on content element', async () => {
    render(
      <>
        <SkipContent selector="#unique-id" focusDelay={1}>
          Aria
        </SkipContent>
        <Section id="unique-id">content</Section>
      </>
    )

    expect(document.activeElement.tagName).toBe('BODY')

    const element = document.querySelector('.dnb-skip-content')

    // 1. make it visible
    fireEvent.keyUp(element.querySelector('button.dnb-sr-only'), {
      key: 'Tab',
      keyCode: 'Tab',
    })

    expect(
      element.querySelector('button.dnb-sr-only')
    ).not.toBeInTheDocument()

    // 2. make focus action
    fireEvent.click(element.querySelector('.dnb-button'))

    await waitFor(() => {
      expect(document.activeElement.tagName).toBe('SECTION')
    })
    expect(element.querySelector('.dnb-button')).not.toBeInTheDocument()
    expect(document.activeElement.classList).toContain(
      'dnb-skip-content__focus'
    )
  })

  it('should hide button when pressing tab key two times (blur second button)', async () => {
    render(
      <>
        <SkipContent selector="#unique-id" focusDelay={1}>
          Aria
        </SkipContent>
        <Section id="unique-id">content</Section>
      </>
    )

    expect(document.activeElement.tagName).toBe('BODY')

    const element = document.querySelector('.dnb-skip-content')

    // 1. make it visible
    fireEvent.keyUp(element.querySelector('button.dnb-sr-only'), {
      key: 'Tab',
      keyCode: 'Tab',
    })

    await waitFor(() => {
      // because of requestAnimationFrame
      expect(document.activeElement.tagName).toBe('BUTTON')
    })

    // 2. blur the event
    fireEvent.blur(element.querySelector('.dnb-button'))

    expect(element.querySelector('.dnb-button')).not.toBeInTheDocument()
    expect(document.activeElement.tagName).toBe('BODY')
  })

  it('should validate axe test', async () => {
    const Component = render(
      <>
        <SkipContent selector="#unique-id">Aria</SkipContent>
        <Section id="unique-id">
          <SkipContent.Return selector="#unique-id">
            return
          </SkipContent.Return>
          content
        </Section>
      </>
    )

    const element = document.querySelector('.dnb-skip-content')

    // make it visible
    fireEvent.keyUp(element.querySelector('button.dnb-sr-only'), {
      key: 'Tab',
      keyCode: 'Tab',
    })

    expect(await axeComponent(Component)).toHaveNoViolations()
  })

  it('scss has to match snapshot', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })
})

describe('SkipContent.Return', () => {
  it('should show button and set focus on content element', async () => {
    render(
      <>
        <SkipContent selector="#unique-id" focusDelay={1}>
          Aria
        </SkipContent>
        <Section id="unique-id">
          <SkipContent.Return selector="#unique-id" focusDelay={1}>
            return
          </SkipContent.Return>
          content
        </Section>
      </>
    )

    expect(document.activeElement.tagName).toBe('BODY')

    const element = document.querySelector('.dnb-skip-content')

    // 1. make it visible
    fireEvent.keyUp(element.querySelector('button.dnb-sr-only'), {
      key: 'Tab',
      keyCode: 'Tab',
    })

    expect(
      element.querySelector('button.dnb-sr-only')
    ).not.toBeInTheDocument()

    // 2. make focus action
    fireEvent.click(element.querySelector('.dnb-button'))

    await waitFor(() => {
      expect(document.activeElement.tagName).toBe('SECTION')
      expect(element.querySelector('.dnb-button')).not.toBeInTheDocument()
      expect(document.activeElement.classList).toContain(
        'dnb-skip-content__focus'
      )
    })

    const section = document.querySelector('#unique-id')

    // 3. make "return" visible
    fireEvent.keyUp(section.querySelector('button.dnb-sr-only'), {
      key: 'Tab',
      keyCode: 'Tab',
    })

    expect(
      section.querySelector('button.dnb-sr-only')
    ).not.toBeInTheDocument()

    // 4. make focus action
    fireEvent.click(section.querySelector('.dnb-button'))

    await waitFor(() => {
      expect(document.activeElement.tagName).toBe('BUTTON')
      expect(section.querySelector('.dnb-button')).not.toBeInTheDocument()
      expect(
        // First parent is HeightAnimation, second is span.dnb-skip-content
        document.activeElement.parentElement.parentElement.getAttribute(
          'id'
        )
      ).toBe('unique-id--alias')
    })
  })
})
