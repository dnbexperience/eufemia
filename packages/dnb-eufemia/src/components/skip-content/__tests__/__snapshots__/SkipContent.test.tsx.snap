// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipContent scss has to match snapshot 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Icon component
*
*/
/*
 * Utilities
 */
.dnb-icon {
  display: inline-block;
  vertical-align: middle;
  font-size: 1rem;
  line-height: 1rem;
  color: inherit;
  width: 1em;
  height: 1em;
}
.dnb-icon img, .dnb-core-style .dnb-icon img,
.dnb-icon svg,
.dnb-core-style .dnb-icon svg {
  width: inherit;
  height: inherit;
  shape-rendering: geometricprecision;
  vertical-align: top;
}
.dnb-icon svg[width="100%"] {
  width: inherit;
}
.dnb-icon svg[height="100%"] {
  height: inherit;
}
.dnb-icon--inherit-color svg:not([fill]), .dnb-icon--inherit-color svg [fill] {
  fill: currentcolor;
}
.dnb-icon--inherit-color svg [stroke] {
  stroke: currentcolor;
}
.dnb-icon--small {
  font-size: 0.75rem;
}
.dnb-icon--default {
  font-size: 1rem;
}
.dnb-icon--medium {
  font-size: 1.5rem;
}
.dnb-icon--large {
  font-size: 2rem;
}
.dnb-icon--x-large {
  font-size: 2.5rem;
}
.dnb-icon--xx-large {
  font-size: 3rem;
}
.dnb-icon--custom-size {
  width: auto;
  height: auto;
}
.dnb-icon--auto {
  font-size: 1em;
}
.dnb-icon--auto > .dnb-icon--wrapper {
  display: inline-flex;
  align-items: center;
  justify-content: center;
}
h1 > .dnb-icon, h2 > .dnb-icon, h3 > .dnb-icon, h4 > .dnb-icon, h5 > .dnb-icon, h6 > .dnb-icon {
  vertical-align: middle;
}
p > .dnb-icon {
  vertical-align: inherit;
}
.dnb-icon.dnb-skeleton {
  color: var(--skeleton-color) !important;
}
.dnb-icon.dnb-skeleton::before, .dnb-icon.dnb-skeleton::after {
  content: none !important;
}

/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Tooltip component
*
*/
/*
 * Utilities
 */
.dnb-tooltip {
  --tooltip-color: white;
  --tooltip-background-color: black;
  --tooltip-border-color: 1px solid black;
  --tooltip-border-radius: 1rem;
  position: absolute;
  z-index: 3100;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0 1rem;
  color: var(--tooltip-color);
  background-color: var(--tooltip-background-color);
  border-radius: var(--tooltip-border-radius);
  box-shadow: var(--shadow-default);
  opacity: 0;
  visibility: hidden;
  transition: opacity 200ms var(--easing-default);
}
.dnb-tooltip--large {
  padding: 0.25rem 1rem;
}
.dnb-tooltip--active {
  visibility: visible;
  /*
    because of the first "show" we also use animation
    also, use forwards because of the usage of visibility
  */
  animation: show-tooltip 200ms var(--easing-default) forwards;
}
html[data-visual-test] .dnb-tooltip--active, .dnb-tooltip--active.dnb-tooltip--no-animation {
  animation: show-tooltip 1ms var(--easing-default) forwards;
}
.dnb-tooltip--hide {
  visibility: visible;
  animation: hide-tooltip 200ms var(--easing-default) forwards;
}
.dnb-tooltip--hide.dnb-tooltip--no-animation {
  animation: hide-tooltip 1ms linear forwards;
}
.dnb-tooltip--fixed {
  position: fixed;
}
html[data-visual-test] .dnb-tooltip--hide {
  animation: hide-tooltip 1ms linear 1s forwards;
}
.dnb-tooltip__portal {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
}
.dnb-tooltip__content {
  min-width: 2rem;
  min-height: 1.5rem;
  padding: 0;
}
.dnb-tooltip__arrow {
  position: absolute;
  pointer-events: none;
  margin: 0;
  width: 1rem;
  height: 0.5rem;
  overflow: hidden;
}
.dnb-tooltip__arrow::before {
  content: "";
  position: absolute;
  bottom: 0;
  left: 0;
  width: 1rem;
  height: 1rem;
  background-color: var(--tooltip-background-color);
  border: var(--tooltip-border-color);
  transform: translateY(70%) rotate(45deg);
}
.dnb-tooltip__arrow__position--bottom {
  top: -0.5rem;
}
.dnb-tooltip__arrow__position--top {
  bottom: -0.5rem;
  transform: rotate(180deg);
}
.dnb-tooltip__arrow__position--left {
  right: -0.75rem;
  margin-right: 3px;
  transform: rotate(90deg);
}
.dnb-tooltip__arrow__position--right {
  left: -0.75rem;
  margin-left: 3px;
  transform: rotate(270deg);
}
.dnb-tooltip__arrow__arrow--left {
  align-self: flex-start;
}
.dnb-tooltip__arrow__arrow--right {
  align-self: flex-end;
}

@keyframes show-tooltip {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes hide-tooltip {
  from {
    opacity: 1;
  }
  to {
    opacity: 0;
    visibility: hidden;
  }
}
/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * FormStatus component
 *
 */
/*
 * Utilities
 */
.dnb-form-status {
  --form-status-radius: 0.25rem;
  display: flex;
  opacity: 1;
  transition: height 400ms var(--easing-default), opacity 400ms var(--easing-default), margin 400ms var(--easing-default), padding 400ms var(--easing-default);
}
.dnb-form-status--hidden {
  will-change: height, opacity, margin, padding;
  width: 0;
  height: 0;
  opacity: 0;
}
.dnb-form-status--is-animating {
  overflow: hidden;
  width: auto;
}
.dnb-form-status--disappear, .dnb-form-status--hidden {
  margin: 0 !important;
  padding: 0 !important;
}
.dnb-form-status__shell {
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
  min-width: inherit;
  border-radius: var(--form-status-radius);
}
.dnb-form-status__text {
  padding: 0.625rem 1rem;
  cursor: text;
  color: inherit;
  line-height: var(--line-height-small);
  font-size: var(--font-size-small);
  white-space: normal;
}
button .dnb-form-status__text {
  cursor: inherit;
}
.dnb-form-status__text .dnb-anchor {
  font-size: inherit;
}
.dnb-icon + .dnb-form-status__text {
  padding-left: 0.5rem;
}
.dnb-form-status__shell > .dnb-icon {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0.3333333em 0.3333333em 0.3333333em 0.6666666em;
}
.dnb-form-status__size--large .dnb-form-status__text {
  padding-top: 1.125rem;
  padding-bottom: 1.125rem;
}
.dnb-form-status__size--large .dnb-form-status__shell > .dnb-icon {
  margin-top: 0.6666666em;
  margin-bottom: 0.6666666em;
}
.dnb-form-status--stretch {
  flex-grow: 1;
}
.dnb-form-status--stretch .dnb-form-status__shell {
  width: 100%;
}
.dnb-form-status--stretch .dnb-form-status__text {
  max-width: 47rem;
}
.dnb-form-status[hidden] {
  display: none;
}
.dnb-form-status--no-animation, html[data-visual-test] .dnb-form-status {
  transition-duration: 1ms !important;
}

/*
* Button component
*
*/
/*
 * Utilities
 */
:root {
  --button-icon-size: 1rem;
  --button-height: 2.5rem;
  --button-border-radius: calc(var(--button-height) / 2);
}

.dnb-button {
  --button-font-size: var(--font-size-basis);
  --button-font-size-small: var(--font-size-small);
  --button-width: 2.5rem;
  --button-width--small: 1.5rem;
  --button-height--small: 1.5rem;
  --button-width--medium: 2rem;
  --button-height--medium: 2rem;
  --button-width--large: 3rem;
  --button-height--large: 3rem;
  --button-border-width: 0.0625rem;
  --button-border-width--hover: 0.1875rem;
  --button-border-radius--small: calc(var(--button-height--small) / 2);
  --button-border-radius--medium: calc(var(--button-height--medium) / 2);
  --button-border-radius--large: calc(var(--button-height--large) / 2);
  position: relative;
  user-select: none;
  /* stylelint-disable-next-line */
  -webkit-user-select: none;
  cursor: pointer;
  white-space: nowrap;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: var(--button-width);
  height: auto;
  padding: 0;
  border: none;
  border-radius: var(--button-border-radius);
  text-align: left;
  font-size: var(--font-size-small);
  text-decoration: none;
  box-shadow: none;
  /* stylelint-disable-next-line */
}
.dnb-button--wrap {
  overflow-wrap: break-word;
  white-space: normal;
}
.dnb-button, .dnb-core-style .dnb-button {
  line-height: var(--button-height);
}
.dnb-button__text {
  margin: 0.5rem 0;
  font-size: var(--button-font-size);
  line-height: var(--line-height-basis);
  color: inherit;
  transform: translateY(-0.03125rem);
}
.dnb-button__text [data-os=linux] {
  transform: translateY(-0.035rem);
}
.dnb-button__alignment {
  display: inline-block;
  width: 0;
}
.dnb-button__bounding {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  transform: scale(1.1, 1.4);
  background-color: transparent;
  border-radius: var(--button-border-radius);
}
.dnb-button--has-text {
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}
.dnb-button--size-small {
  width: var(--button-width--small);
  font-size: var(--button-font-size-small);
  border-radius: var(--button-border-radius--small);
}
.dnb-button--size-small, .dnb-core-style .dnb-button--size-small {
  line-height: var(--button-height--small);
}
.dnb-button--size-small .dnb-button__text {
  margin: 0;
}
.dnb-button--has-text.dnb-button--size-small {
  padding-left: 1rem;
  padding-right: 1rem;
}
.dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-small {
  padding-left: 0.5rem;
}
.dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-small {
  padding-right: 0.5rem;
}
.dnb-button--size-medium {
  width: var(--button-width--medium);
  border-radius: var(--button-border-radius--medium);
}
.dnb-button--size-medium, .dnb-core-style .dnb-button--size-medium {
  line-height: var(--button-height--medium);
}
.dnb-button--size-medium .dnb-button__text {
  margin: 0;
}
.dnb-button--has-text.dnb-button--size-medium {
  padding-left: 1rem;
  padding-right: 1rem;
}
.dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-medium {
  padding-left: 0.5rem;
}
.dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-medium {
  padding-right: 0.5rem;
}
.dnb-button--size-large {
  width: var(--button-width--large);
  border-radius: var(--button-border-radius--large);
}
.dnb-button--size-large, .dnb-core-style .dnb-button--size-large {
  line-height: var(--button-height--large);
}
.dnb-button--has-text.dnb-button--size-large {
  padding-left: 2rem;
  padding-right: 2rem;
}
.dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-large {
  padding-left: 1rem;
}
.dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-large {
  padding-right: 1rem;
}
.dnb-button--has-text {
  width: auto;
}
.dnb-button--has-text .dnb-button__icon {
  margin: 0 calc(var(--button-icon-size) / 2);
}
.dnb-button--has-text.dnb-button--icon-position-left {
  padding-left: 0.5rem;
}
.dnb-button--has-text.dnb-button--icon-position-right {
  padding-right: 0.5rem;
}
.dnb-button--has-text.dnb-button--has-icon .dnb-button__icon {
  order: 2;
}
.dnb-button--has-text.dnb-button--has-icon .dnb-button__text {
  order: 1;
}
.dnb-button:not(.dnb-button--has-text) .dnb-button__icon {
  width: inherit;
}
.dnb-button__icon.dnb-icon svg:not([width]):not([height]) {
  width: var(--button-icon-size);
  height: var(--button-icon-size);
}
[href] > .dnb-button__icon.dnb-icon {
  line-height: var(--button-font-size);
}
.dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__icon, .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-top .dnb-button__icon {
  order: 1;
}
.dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__text, .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-top .dnb-button__text {
  order: 2;
}
.dnb-button--stretch {
  width: 100%;
}
.dnb-button--reset {
  margin: 0;
  padding: 0;
  width: auto;
  height: auto;
  overflow: visible;
  border: none;
  border-radius: 0;
  background-color: transparent;
  appearance: none;
  box-shadow: none;
  color: inherit;
  font: inherit;
  text-align: inherit;
  line-height: inherit;
}
html:not([data-whatintent=touch]) .dnb-button--reset:hover[disabled] {
  cursor: not-allowed;
}
html:not([data-whatintent=touch]) .dnb-button--reset:hover:not([disabled]) {
  box-shadow: none;
  border: none;
}
.dnb-button--reset:not([disabled]):focus, .dnb-button--reset:not([disabled]):active {
  outline: none;
}
html[data-whatinput=keyboard] .dnb-button--reset:not([disabled]):focus, html[data-whatinput=keyboard] .dnb-button--reset:not([disabled]):active {
  --border-color: var(--focus-ring-color);
  --border-width: var(--focus-ring-width);
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
html[data-whatinput=mouse] .dnb-button--reset:not([disabled]):focus, html[data-whatinput=mouse] .dnb-button--reset:not([disabled]):active {
  box-shadow: none;
  color: inherit;
  border: none;
}
.dnb-button[type=button], .dnb-button[type=reset], .dnb-button[type=submit] {
  appearance: none;
  /* stylelint-disable-next-line */
  -moz-appearance: none;
  /* stylelint-disable-next-line */
  -webkit-appearance: none;
}
.dnb-button[disabled] {
  cursor: not-allowed;
  outline: none;
}
.dnb-form-row--vertical .dnb-form-row__content > .dnb-button {
  align-self: flex-start;
}
.dnb-form-row--horizontal .dnb-form-row__content .dnb-button__text {
  white-space: nowrap;
}
.dnb-button + .dnb-form-status {
  margin-top: 0.5rem;
}

/* Firefox includes a hidden border which messes up button dimensions */
button.dnb-button::-moz-focus-inner {
  border: none;
}

/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
 * Section component
 *
 */
/*
 * Utilities
 */
.dnb-section {
  position: relative;
  display: flow-root;
  --breakout--on: visible;
  --breakout--off: hidden;
  --rounded-corner--value: 0.75rem;
  --outline: 0 0 0 1px var(--outline-color, transparent);
  --top: 0;
  --left: -100vw;
  --width: 100vw;
  --height: 100%;
  box-shadow: var(--outline, none);
  color: var(--text-color, black);
  border-radius: var(--rounded-corner, 0);
}
.dnb-section[style*="--background-color"] {
  background-color: var(--background-color, white);
}
.dnb-section::after {
  content: "";
  visibility: var(--breakout, hidden);
  position: absolute;
  z-index: -15;
  left: var(--left);
  top: var(--top);
  width: var(--width);
  height: var(--height);
  color: var(--background-color, currentcolor);
  background-color: currentcolor;
  --box-shadow: 99vw 0 0 0 currentcolor, 198vw 0 0 0 currentcolor,
    297vw 0 0 0 currentcolor, 396vw 0 0 0 currentcolor;
  box-shadow: var(--box-shadow);
}
.dnb-section:not([style*="--breakout"]) {
  --breakout: var(--breakout--on);
}
@media screen and (max-width: 60em) {
  .dnb-section {
    --breakout: var(--breakout--small, var(--breakout--fallback));
    --background-color--value: var(--background-color--small);
    --text-color--value: var(--text-color--small);
    --outline-color: var(--outline-color--small);
    --rounded-corner: var(
      --rounded-corner--small,
      var(--rounded-corner--fallback)
    );
  }
}
@media screen and (max-width: 59.9375em) and (min-width: 40.0625em) {
  .dnb-section {
    --breakout: var(--breakout--medium, var(--breakout--fallback));
    --background-color--value: var(--background-color--medium);
    --text-color--value: var(--text-color--medium);
    --outline-color: var(--outline-color--medium);
    --rounded-corner: var(
      --rounded-corner--medium,
      var(--rounded-corner--fallback)
    );
  }
}
@media screen and (min-width: 60em) {
  .dnb-section {
    --breakout: var(--breakout--large, var(--breakout--fallback));
    --background-color--value: var(--background-color--large);
    --text-color--value: var(--text-color--large);
    --outline-color: var(--outline-color--large);
    --rounded-corner: var(
      --rounded-corner--large,
      var(--rounded-corner--fallback)
    );
  }
}
.dnb-section .dnb-section::after {
  z-index: -14;
}
.dnb-section .dnb-section .dnb-section::after {
  z-index: -13;
}
.dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -12;
}
.dnb-section .dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -11;
}
.dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -10;
}
.dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -9;
}
.dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -8;
}
.dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section .dnb-section::after {
  z-index: -7;
}
.dnb-section--spacing-x-small {
  padding-top: var(--spacing-x-small);
  padding-bottom: var(--spacing-x-small);
}
.dnb-section--spacing-small {
  padding-top: var(--spacing-small);
  padding-bottom: var(--spacing-small);
}
.dnb-section--spacing-medium {
  padding-top: var(--spacing-medium);
  padding-bottom: var(--spacing-medium);
}
.dnb-section--spacing, .dnb-section--spacing-large {
  padding-top: var(--spacing-large);
  padding-bottom: var(--spacing-large);
}
.dnb-section--spacing-x-large {
  padding-top: var(--spacing-x-large);
  padding-bottom: var(--spacing-x-large);
}
.dnb-section--spacing-xx-large {
  padding-top: var(--spacing-xx-large);
  padding-bottom: var(--spacing-xx-large);
}

/*
* SkipContent component
*
*/
/*
 * Utilities
 */
.dnb-skip-content--visible {
  display: flex;
}
.dnb-skip-content .dnb-button {
  margin: 1rem 0;
}
.dnb-skip-content__focus:focus {
  position: relative;
}
.dnb-skip-content__focus:focus::before {
  content: "";
  position: absolute;
  inset: var(--focus-ring-width);
  outline: none;
}
html[data-whatinput=keyboard] .dnb-skip-content__focus:focus::before {
  --border-color: var(--focus-ring-color);
  --border-width: var(--focus-ring-width);
  box-shadow: 0 0 0 var(--border-width) var(--border-color);
  border-color: transparent;
}
.dnb-skip-content__return:not(.dnb-skip-content__return--active) button.dnb-sr-only {
  display: none;
}"
`;
